// @nuxt/hints/dist/module.mjs
import { addDevServerHandler, defineNuxtModule, createResolver, addPlugin, addBuildPlugin, addComponent, addServerPlugin } from '@nuxt/kit';
import { existsSync } from 'node:fs';
import { eventHandler, proxyRequest } from 'h3';
import MagicString from 'magic-string';
import { createUnplugin } from 'unplugin';

const DEVTOOLS_UI_ROUTE = "/__nuxt-hints";
const DEVTOOLS_UI_LOCAL_PORT = 3300;
function setupDevToolsUI(nuxt, resolver) {
  const clientPath = resolver.resolve("./client");
  const isProductionBuild = existsSync(clientPath);
  if (isProductionBuild) {
    nuxt.hook("vite:serverCreated", async (server) => {
      const sirv = await import('sirv').then((r) => r.default || r);
      server.middlewares.use(
        DEVTOOLS_UI_ROUTE,
        sirv(clientPath, { dev: true, single: true })
      );
    });
  } else {
    addDevServerHandler({
      route: DEVTOOLS_UI_ROUTE,
      handler: eventHandler((e) => {
        return proxyRequest(e, "http://localhost:" + DEVTOOLS_UI_LOCAL_PORT + DEVTOOLS_UI_ROUTE + e.path);
      })
    });
  }
  nuxt.hook("devtools:customTabs", (tabs) => {
    tabs.push({
      name: "hints",
      title: "Hints",
      icon: "carbon:idea",
      category: "analyze",
      view: {
        type: "iframe",
        src: DEVTOOLS_UI_ROUTE
      }
    });
  });
}

const InjectHydrationPlugin = createUnplugin(() => {
  return {
    name: "@nuxt/hints:inject-hydration-check",
    enforce: "pre",
    transformInclude(id) {
      return id.endsWith(".vue") && !id.includes("node_modules");
    },
    transform(code) {
      const m = new MagicString(code);
      const re = /<script\s+setup[^>]*>/g;
      const match = re.exec(code);
      if (!match) {
        return code;
      }
      m.appendRight(
        match.index + match[0].length,
        `
import { useHydrationCheck } from '@nuxt/hints/runtime/composables/hydration'
useHydrationCheck();`
      );
      return {
        code: m.toString(),
        map: m.generateMap({ hires: true })
      };
    }
  };
});

const module = defineNuxtModule({
  meta: {
    name: "@nuxt/hints",
    configKey: "hints"
  },
  defaults: {
    devtools: true
  },
  setup(options, nuxt) {
    if (!nuxt.options.dev) {
      return;
    }
    const resolver = createResolver(import.meta.url);
    addPlugin(resolver.resolve("./runtime/plugins/web-vitals/plugin.client"));
    addPlugin(resolver.resolve("./runtime/plugins/hydration/plugin.client"));
    addBuildPlugin(InjectHydrationPlugin);
    addComponent({
      name: "NuxtIsland",
      filePath: resolver.resolve("./runtime/components/nuxt-island"),
      priority: 1e3
    });
    addPlugin(resolver.resolve("./runtime/plugins/third-party-scripts/plugin.client"));
    addServerPlugin(resolver.resolve("./runtime/plugins/third-party-scripts/nitro.plugin"));
    nuxt.hook("prepare:types", ({ references }) => {
      references.push({
        types: resolver.resolve("./runtime/types.d.ts")
      });
    });
    if (options.devtools) {
      setupDevToolsUI(nuxt, resolver);
      addPlugin(resolver.resolve("./runtime/plugins/vue-tracer-state.client"));
    }
  }
});

export { module as default };
